sdk-rs/src/store/adb.rs

```diff
@@ -0,0 +1,192 @@
+pub struct GetResultPayload {
+    /// The requested key's value.
+    #[serde_as(as = "Base64")]
+    pub value: Vec<u8>,
+
+    /// The position of the value in the MMR.
+    pub position: u64,
+
+    /// The raw proof data for the value that can be verified against the database's root at the
+    /// state corresponding to the provided MMR size.
+    #[serde_as(as = "Base64")]
+    pub proof_data: Vec<u8>,
+}
+
+impl Client {
+    pub fn new(client: SdkClient, parent_url: &str) -> Self {
+        Self {
+            client,
+            base_url: format!("{parent_url}{PATH}"),
+        }
+    }
+
+    pub async fn get_and_verify_proof(
+        &self,
+        root: [u8; 32],
+        key: &[u8],
+        mmr_size: u64,
+    ) -> Result<Option<Vec<u8>>, Error> {
+        let res = self.get(key, mmr_size).await?;
+        let Some(payload) = res else {
+            return Ok(None);
+        };
+
+        if payload.proof_data.len() % <<Sha256 as Hasher>::Digest as FixedSize>::SIZE != 0 {
+            return Err(Error::BadResponse);
+        }
+
+        let mut digests = Vec::with_capacity(
+            payload.proof_data.len() / <<Sha256 as Hasher>::Digest as FixedSize>::SIZE,
+        );
+        // Convert bytes to digests
+        for chunk in payload
+            .proof_data
+            .chunks(<<Sha256 as Hasher>::Digest as FixedSize>::SIZE)
+        {
+            let mut digest = [0u8; 32];
+            digest.copy_from_slice(chunk);
+            digests.push(digest.into());
+        }
+
+        let proof: Proof<<Sha256 as Hasher>::Digest> = Proof {
+            size: mmr_size,
+            digests,
+        };
+
+        let mut hasher = Standard::<Sha256>::new();
+        let root_digest: Sha256Digest = root.into();
+        let _root = proof.verify_range_inclusion(
+            &mut hasher,
+            &[payload.value.clone()],
+            payload.position,
+            &root_digest,
+        );
+
+        Ok(Some(payload.value))
+    }
+
+    /// Retrieves a value from the store by its key, along with a proof that should verify against
+    /// the database's root corresponding to the provided MMR size.
+    ///
+    /// - Returns `Ok(None)` if the key does not exit.
+    /// - Returns [Error::ServerBehind] if the server's state is not sufficiently up-to-date to
+    ///   handle a query for the provided size.
+    pub async fn get(&self, key: &[u8], mmr_size: u64) -> Result<Option<GetResultPayload>, Error> {
+        let (url, headers) = self.get_request(key, mmr_size, HeaderMap::new());
+
+        let res = self.client.get(url, headers).await?;
+
+        Self::get_handle_response(res).await
+    }
+
+    fn get_request(
+        &self,
+        key: &[u8],
+        mmr_size: u64,
+        mut headers: HeaderMap,
+    ) -> (String, HeaderMap) {
+        let key_b64 = general_purpose::STANDARD.encode(key);
+        let mut url = self.base_url.clone();
+        url.push_str(&format!("?key={key_b64}&size={mmr_size}"));
```

Comment:
IIRC there is a way to attach query params here without doing this manual formatting (I may have committed this sin elsewhere)

Reply:
yep I was just following your kv.rs example 😄 

Should I change everything over to use the rust url::Url?

Reply:
Changed to use url::Url